File: src/ns/browser-item.ts
import { Ableton } from "..";
import { Namespace } from ".";

export interface GettableProperties {
  children: RawBrowserItem[];
  is_device: boolean;
  is_folder: boolean;
  is_loadable: boolean;
  is_selected: boolean;
  name: string;
  source: string;
  uri: string;
}

export interface TransformedProperties {
  children: BrowserItem[];
}

export interface SettableProperties {}

export interface ObservableProperties {}

export interface RawBrowserItem {
  readonly id: string;
  readonly children: RawBrowserItem[];
  readonly name: string;
  readonly is_loadable: boolean;
  readonly is_selected: boolean;
  readonly is_device: boolean;
  readonly is_folder: boolean;
  readonly source: string;
  readonly uri: string;
}

export class BrowserItem extends Namespace<
  GettableProperties,
  TransformedProperties,
  SettableProperties,
  ObservableProperties
> {
  constructor(
    ableton: Ableton,
    public raw: RawBrowserItem,
  ) {
    super(ableton, "browser-item", raw.id);
    this.transformers = {
      children: (children) => children.map((c) => new BrowserItem(ableton, c)),
    };

    this.cachedProps = {
      children: true,
      is_device: true,
      is_folder: true,
      is_loadable: false,
      is_selected: false,
      name: true,
      source: true,
      uri: true,
    };
  }
}

File: midi-script/BrowserItem.py
from __future__ import absolute_import
from .Interface import Interface


class BrowserItem(Interface):
    @staticmethod
    def serialize_browser_item(browser_item):
        if browser_item is None:
            return None
        browser_item_id = Interface.save_obj(browser_item)
        return {
            "id": browser_item_id,
            "name": browser_item.name,
            "is_loadable": browser_item.is_loadable,
            "is_selected": browser_item.is_selected,
            "is_device": browser_item.is_device,
            "is_folder": browser_item.is_folder,
            "source": browser_item.source,
            "uri": browser_item.uri,
        }

    def __init__(self, c_instance, socket):
        super(BrowserItem, self).__init__(c_instance, socket)

    def get_children(self, ns):
        return map(BrowserItem.serialize_browser_item, ns.children)
